### PR TITLE
Fix Font Hinting

### LINKED ISSUE #293: Small font rendering really unclear (blurry, missing pixels etc)
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have verified that I am running the latest version of Fonts
- [X] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [X] I have searched [open](https://github.com/SixLabors/Fonts/issues) and [closed](https://github.com/SixLabors/Fonts/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

Rendering small fonts is unreadable with alpha blending enabled, due to the really imprecise, blurry way they're being rendered:

ImageSharp (default antialiasing settings):
image

GDI+ (clear type):
image

This is with Segoe UI at 10 pixels. The vertical lines especially aren't clear, and with alpha blending enabled they just disappear. 

I also tried disabling antialiasing in the ImageSharp render, but that makes it even more obvious how imprecise it is:
image

I also tried hinting, which is supposed to help with this, but it doesn't seem to work at all:
image

For Open Sans it's even less readable, and hinting just produces long vertical lines:
image

Is there some set of options I can enable to make small font rendering clear?

### Steps to Reproduce
I have a sample repo that I used to generate those images at https://github.com/myblindy/font_test

### System Configuration

- Fonts version: 1.0.0-beta18
- Other Six Labors packages and versions: ImageSharp 2.1.3, ImageSharp.Drawing 1.0.0-beta15
- Environment (Operating system, version and so on): Windows 10
- .NET Framework version: .Net 6
- Additional information: N/A

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Fixes #293 and #294

We were passing the multiplied pixel size when we shouldn't have. Some pre-cleartype fonts don't seem to work all that well so I've internally disabled hinting for them for now.

### CODE REVIEW

File: src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs [hunk 1]
```diff
@@ -21,11 +21,31 @@ namespace SixLabors.Fonts
...
+        /// These common fonts can be hinted by the engine only in XY mode and only if we do not scale the x-dimension.
+        /// even then the output is worse than the non-hinted version due to thinning of certain strokes.
+        /// <see href="http://agg.sourceforge.net/antigrain.com/research/font_rasterization/"/>.
+        /// </para>
+        /// <para>
+        /// We'll sacrifice hinting for these fonts to allow the engine to work better for modern fonts and avoid potentially raised issues.
+        /// </para>
+        /// </summary>
+        private static readonly string[] BadHintList = { "Arial", "Times New Roman", "Tahoma", "Palatino Linotype", "Segoe UI" };
+        private bool? shouldHint;
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: There's a LOT of knowledge in this document, unfortunately I couldn't find a way to get these fonts to hint well. They're all older ones that have known issues with hinting so to save effort I've disabled hinting of them for now.

http://rastertragedy.com/RTRCh4.htm#Sec1

Every single modern font I tested seems to work really well.

Comment (tocsoft) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: is there any metadata in the font we can use to determine if we should enable hinting? some created date maybe or so versioning flag difference? anything to save just hard coding a set of font names.

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: Unfortunately not. 😔

There's some guidance for how to actually mitigate the issues I saw but I spent a few days and couldn't translate it into code. Nor could I find reference in FreeType that provided a complete workaround matching that guidance.

> Accordingly I was looking for an approach that combines the advantages of overscaling before constraint application with the “safety” of overscaling after constraint application. To do so I “jury rigged” the rasterizer to dynamically re-interpret all rounding instructions to mean “round to sample boundary” if the rendering method is ClearType and if the direction of the projection vector p is not parallel to the long side of the LCD sub-pixels, else to keep the meaning “round to pixel boundary” as before:
> 
> - Apply contextually interpreted constraints (“hints”) to outlines
> - Overscale constrained outlines in x-direction by 6x
> - Sample overscaled constrained outlines into overscaled bitmaps
> - Apply anti-aliasing filter to downsample overscaled bitmaps
> 
> Semantically, the two interpretations of rounding instructions are most similar. In fact, if you consider full-pixel bi-level rendering (colloquially “black-and-white” rendering) as “oversampling” at the rate of one sample per pixel, then the two interpretations are identical. Practically, this approach eliminates the “explosions” and it overcomes the absence of a TrueType instruction that would permit “finer grained” rounding than to the nearest 1/2 of a pixel (cf [4.3.0](http://rastertragedy.com/RTRCh4.htm#Sec30)).

File: src/SixLabors.Fonts/Tables/TrueType/Hinting/TrueTypeInterpreter.cs [hunk 2]
```diff
@@ -1398,7 +1399,7 @@ private void ShiftPoints(Vector2 displacement)
             state.Loop = 1;
         }
 
-        private void MovePoint(Zone zone, int index, float distance)
+        private void MovePoint(Zone zone, int index, float distance, [CallerMemberName] string member = null)
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Tables/TrueType/Hinting/TrueTypeInterpreter.cs:1402]: Ah yes. That was me trying to figure out where exactly a glyph  explosion was getting triggered.

File: src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: I do wonder whether we just just try tagging the chap who built the original interpreter to see if he knows how?

File: src/SixLabors.Fonts/Buffer{T}.cs [hunk 3]
```diff
@@ -26,21 +27,37 @@ public Buffer(int length)
...
+        /// </summary>
+        /// <returns>The <see cref="ArraySlice{T}"/>.</returns>
+        public ArraySlice<T> DangerousGetSlice()
+        {
+            if (this.buffer is null)
+            {
+                ThrowObjectDisposedException();
+            }
+
+            return new ArraySlice<T>(Unsafe.As<T[]>(this.buffer), 0, this.length);
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/Buffer{T}.cs:50]: This is dangerous. Will look for a workaround.

File: src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: I hate to do this and apologies in advance but @MikePopoloski would you have any insight into how to implement some of these workarounds?

Comment (MikePopoloski) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: Oh hello. No sorry, I haven't thought much about font rendering in like 7 years. Glad someone is making use of the interpreter though, it's a neat bit of code and there aren't (or at least weren't when I built it) too many open source implementations of a TT interpreter.

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/StreamFontMetrics.TrueType.cs:36]: Thanks for replying! That was a bit of a hit and hope tagging you like that. 

I think there's very few people in the world who actually understand the TT interpreter. I've yet to find an implementation other than yours in the .NET ecosystem and elsewhere there just appears to be using Freetype.

I'll persevere though and continue to do my own research. Fongers crossed I can figure it all out. It's an interesting challenge!

File: src/SixLabors.Fonts/Buffer{T}.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Buffer{T}.cs:52]: @antonfirsov Do you have any ideas here?

I never let anything access the buffer outside of the range, but it still doesn't feel right that I have this unsafe cast array that returns the length in the number of bytes.

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Buffer{T}.cs:52]: @antonfirsov sorry to bump. I'm just out of ideas without rewriting the interpreter to work with spans

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Buffer{T}.cs]: Decided to take the perf hit and go safe for now.

**Comment by JimBobSquarePants** [comment]
@tocsoft I've cracked it! 🎉

I had a good dig through the FreeType interpreter code and accompanying docs and figured out what they did to get their interpreter working well with legacy fonts. Basically, they never move anything horizontally when hinting at all which fixes most issues then there's a couple of additional tweaks for legacy fonts. No need for lists or targeted hacks - it just works!

Here's Tahoma (_This was really bad before with odd offsetting of individual glyphs._)
 
`HintingMode.HintXY`
_six_tahoma_hintxy_

`HintingMode.HintY`
_six_tahoma_hinty_

`HintingMode.None`
_six_tahoma_none_

`HintingMode.HintXY` vs `HintingMode.None` (_You can clearly see the hinting benefit in the 'o' here_)
image

And here's Open Sans

`HintingMode.HintXY` (_Both hint modes Y and XY produce identical results with this font_)
_six_open_sans_hintxy_

`HintingMode.None`
_six_open_sans_none_

`HintingMode.HintXY` vs `HintingMode.None`  (_You can clearly see the hinting benefit in the 'o' here_)
image

For most fonts there really miniscule difference between `HintingMode.HintXY` and `HintingMode.HintY`. There are some notable improvements in Times New Roman when using `HintingMode.HintY` in that the over-thinning of stems does not happen. I'm thinking of reworking the enum to rename `HintingMode.HintY` as `HintingMode.Compatibility` moving it to the third position as there's additional overheads (_scaling of control point x-values before and after hinting_) which, while SIMD optimized, you don't really want people using as a first-choice approach.

What do you think?

image

**Review by tocsoft — DISMISSED** [review]
This looks fantastic, congrats on getting it working without the black list.

**Review by tocsoft — APPROVED** [review]

**Pull request merged**